charlotte_core/src/framebuffer/framebuffer.rs

```diff
@@ -0,0 +1,176 @@
+#![no_std]
+
+use core::fmt::write;
+use crate::chars::*;
+
+extern crate limine;
+
+pub struct FrameBufferInfo {
+    address: *mut u32,
+    width: usize,
+    height: usize,
+    pitch: usize,
+    bpp: usize,
+}
+
+#[derive(Copy, Clone)]
+pub struct Point {
+    pub x: isize,
+    pub y: isize,
+}
+
+
+fn blend_colors(foreground: u32, background: u32, blend_factor: u8) -> u32 {
+    let fg_ratio = blend_factor as u32;
+    let bg_ratio = 255 - fg_ratio as u32;
+
+    let r = (((foreground >> 16) & 0xFF) * fg_ratio + ((background >> 16) & 0xFF) * bg_ratio) / 255;
+    let g = (((foreground >> 8) & 0xFF) * fg_ratio + ((background >> 8) & 0xFF) * bg_ratio) / 255;
+    let b = ((foreground & 0xFF) * fg_ratio + (background & 0xFF) * bg_ratio) / 255;
+
+    (r << 16) | (g << 8) | b
+}
+
+
+
+impl FrameBufferInfo {
+    pub fn new (framebuffer: &limine::Framebuffer) -> Self {
+        Self {
+            address: framebuffer.address.as_ptr().unwrap() as *mut u32,
+            width: framebuffer.width as usize,
+            height: framebuffer.height as usize,
+            pitch: framebuffer.pitch as usize,
+            bpp: framebuffer.bpp as usize
+        }
+    }
+
+     // Bresenham's line algorithm
+     unsafe fn draw_line(&self, p0: Point, p1: Point, color: u32) {
+        let mut x0 = p0.x;
+        let mut y0 = p0.y;
+        let x1 = p1.x;
+        let y1 = p1.y;
+
+        let dx = isize::abs(x1 - x0);
+        let dy = -isize::abs(y1 - y0);
+        let sx = if x0 < x1 { 1 } else { -1 };
+        let sy = if y0 < y1 { 1 } else { -1 };
+        let mut err = dx + dy; // error value e_xy
+
+        loop {
+            self.draw_pixel(x0 as usize, y0 as usize, color); // Draw the current pixel
+            if x0 == x1 && y0 == y1 { break; }
+            let e2 = 2 * err;
+            if e2 >= dy { // e_xy+e_x > 0
+                err += dy;
+                x0 += sx;
+            }
+            if e2 <= dx { // e_xy+e_y < 0
+                err += dx;
+                y0 += sy;
+            }
+        }
+    }
+
+    pub unsafe fn clear_screen(&self, color: u32) {
+        for y in 0..self.height {
+            for x in 0..self.width {
+                let pixel_offset = y * self.pitch / (self.bpp / 8) +x;
+               *self.address.add(pixel_offset) = color;
+            }
+        }
+    }
+
+    pub unsafe fn draw_pixel(&self, x: usize, y: usize, color: u32) {
+        if x < self.width && y < self.height {
+            let pixel_offset = y * self.pitch / (self.bpp / 8) + x;
+            *self.address.add(pixel_offset) = color;
+        }
+    } 
+
+    pub unsafe fn draw_text(&self, mut x: usize, mut y: usize, text: &str, color: u32) {
+        let start_x = x; // Remember the starting x position to reset to it on new lines
+        for c in text.chars() {
+            if c == ' ' {
+                // Handle spaces by advancing the x position
+                x += FONT_WIDTH + 1; // Plus 1 for a gap between characters
```

Comment:
We could instead create a space character of all 0s and just draw that onto the screen, this works for now though.